homeassistant/helpers/entity_registry.py

```diff
@@ -286,18 +286,33 @@ def async_remove(self, entity_id: str) -> None:
         )
         self.async_schedule_save()
 
-    @callback
-    def async_device_removed(self, event: Event) -> None:
-        """Handle the removal of a device.
+    async def async_device_modified(self, event: Event) -> None:
+        """Handle the removal or update of a device.
 
         Remove entities from the registry that are associated to a device when
         the device is removed.
+
+        Disable entities in the registry that are associated to a device when
+        the device is disabled.
         """
-        if event.data["action"] != "remove":
-            return
-        entities = async_entries_for_device(self, event.data["device_id"])
-        for entity in entities:
-            self.async_remove(entity.entity_id)
+        if event.data["action"] == "remove":
+            entities = async_entries_for_device(self, event.data["device_id"])
+            for entity in entities:
+                self.async_remove(entity.entity_id)
+
+        if event.data["action"] == "update":
+            device_registry = (
+                await self.hass.helpers.device_registry.async_get_registry()
+            )
+            device = device_registry.async_get(event.data["device_id"])
+            if not device.disabled:
+                return
+
+            entities = async_entries_for_device(self, event.data["device_id"])
+            for entity in entities:
+                self.async_update_entity(  # type: ignore
+                    entity.entity_id, disabled_by=device.disabled_by
```

Comment:
If an entity is going to be disabled because disabling a device, it should be `disabled_by="device"`